pydantic/_internal/_fields.py

```diff
@@ -79,3 +90,162 @@ def _update_attrs(self, constraints: dict[str, Any], attrs: set[str] | None = No
+    :param typevars_map: TODO ???
+    :param is_dataclass: whether the class is a dataclass, used to decide about kw_only setting
+    :param dc_kw_only: whether the whole dataclass is kw_only
+    """
+    from ..fields import FieldInfo
+
+    module_name = getattr(cls, '__module__', None)
+    global_ns: dict[str, Any] | None = None
+    if module_name:
+        try:
+            global_ns = sys.modules[module_name].__dict__
+        except KeyError:
+            # happens occasionally, see https://github.com/pydantic/pydantic/issues/2363
+            pass
+
+    # get type hints and raise a PydanticUndefinedAnnotation if any types are undefined
+    try:
+        type_hints = get_type_hints(cls, global_ns, types_namespace, include_extras=True)
+    except NameError as e:
+        try:
+            name = e.name
+        except AttributeError:
+            m = re.search(r".*'(.+?)'", str(e))
+            if m:
+                name = m.group(1)
+            else:
+                # should never happen
+                raise
+        raise PydanticUndefinedAnnotation(name) from e
+
+    # https://docs.python.org/3/howto/annotations.html#accessing-the-annotations-dict-of-an-object-in-python-3-9-and-older
+    # annotations is only used for finding fields in parent classes
+    annotations = cls.__dict__.get('__annotations__', {})
+    fields: dict[str, FieldInfo] = {}
+
+    # FIXME (dataclasses) what are we doing here?
+    # cls_fields: dict[str, FieldInfo] = getattr(cls, 'model_fields', None) or getattr(cls, '__pydantic_fields__', {})
+
+    # currently just used for `init=False` dataclass fields, but could be used more
+    omitted_fields: set[str] | None = getattr(cls, '__pydantic_omitted_fields__', None)
+
+    for ann_name, ann_type in type_hints.items():
+        if ann_name.startswith('_') or is_classvar(ann_type) or (omitted_fields and ann_name in omitted_fields):
+            continue
+
+        # raise a PydanticUndefinedAnnotation if type is undefined
+        if isinstance(ann_type, ForwardRef):
+            raise PydanticUndefinedAnnotation(ann_type.__forward_arg__)
+
+        # FIXME (dataclasses) what are we doing here?
+        pydantic_generic_typevars_map = getattr(cls, '__pydantic_generic_typevars_map__', None)
+        if pydantic_generic_typevars_map:
+            ann_type = replace_types(ann_type, pydantic_generic_typevars_map)
+
+        if ann_type is dataclasses.KW_ONLY:
+            # all field fields will be kw_only
+            dc_kw_only = True
+            continue
+        kw_only = dc_kw_only
+
+        init_var = False
+        if ann_type is dataclasses.InitVar:
+            init_var = True
+            ann_type = Any
+        elif isinstance(ann_type, dataclasses.InitVar):
+            init_var = True
+            ann_type = ann_type.type
+
+        # FIXME (dataclasses) what are we doing here?
+        generic_origin = getattr(cls, '__pydantic_generic_origin__', ())
```

Comment:
Not sure I quite understand you here.

Reply:
If I have

```python
class MyModel(BaseModel, Generic[T]):
    x: T

MyModel[int]
```
It will go through the class building process when you do `MyModel[int]` — and it will create a subclass of `MyModel` that has a field `x: int`, where the parent class had `x: T`. The `generic_origin` stuff starting with the highlighted line here is ensuring that this doesn't produce a `you are shadowing a field` warning/error.

Reply:
makes sense, I've added a comment.